File: cpp/exercises/practice/unit_test_25e5dd90_cef5_4d7d_804e_9e8b585e01ef/task_scheduler/.meta/task_scheduler.cpp
#include "task_scheduler.h"
#include <queue>
#include <vector>
#include <algorithm>

struct RunningTask {
    int finish_time;
    int task_id;
    // For resource freeing, we store the resource requirements as is.
    std::vector<std::pair<int,int>> resource_requirements;
};

struct RunningTaskComparator {
    bool operator()(const RunningTask& a, const RunningTask& b) const {
        return a.finish_time > b.finish_time; // min-heap: the smallest finish_time has highest priority.
    }
};

int scheduleTasks(const std::vector<Task>& tasks, const std::vector<Resource>& resources) {
    int n = tasks.size();
    if(n == 0) return 0;

    // Build dependency graph and compute in-degree.
    std::vector<int> inDegree(n, 0);
    std::vector<std::vector<int>> dependents(n);
    for (int i = 0; i < n; ++i) {
        for (int dep : tasks[i].dependencies) {
            // task i depends on dep, so dep is prerequisite of i.
            dependents[dep].push_back(i);
            inDegree[i]++;
        }
    }

    // Ready task list: tasks with no remaining dependencies.
    std::vector<int> ready;
    for (int i = 0; i < n; ++i) {
        if (inDegree[i] == 0) {
            ready.push_back(i);
        }
    }
    // Sort ready tasks by id to ensure deterministic ordering.
    std::sort(ready.begin(), ready.end());

    // Available resources.
    std::vector<int> available;
    available.reserve(resources.size());
    for (const auto& res : resources) {
        available.push_back(res.capacity);
    }

    // Priority queue for running tasks.
    std::priority_queue<RunningTask, std::vector<RunningTask>, RunningTaskComparator> runningTasks;

    int current_time = 0;
    int finishedCount = 0;
    // To keep track if a task has been scheduled.
    std::vector<bool> scheduled(n, false);

    while (finishedCount < n) {
        bool scheduledSomething = false;
        std::vector<int> remainingReady;
        // Try to schedule tasks in ready if resources are available.
        // We process in order of increasing task id.
        for (int taskId : ready) {
            const Task& task = tasks[taskId];
            bool canSchedule = true;
            for (auto& req : task.resource_requirements) {
                int resId = req.first;
                int qty = req.second;
                if (resId >= static_cast<int>(available.size()) || available[resId] < qty) {
                    canSchedule = false;
                    break;
                }
            }
            if (canSchedule) {
                // Allocate resources.
                for (auto& req : task.resource_requirements) {
                    int resId = req.first;
                    int qty = req.second;
                    available[resId] -= qty;
                }
                // Schedule the task.
                RunningTask rt;
                rt.task_id = taskId;
                rt.finish_time = current_time + task.duration;
                rt.resource_requirements = task.resource_requirements;
                runningTasks.push(rt);
                scheduled[taskId] = true;
                scheduledSomething = true;
            } else {
                // Keep task in ready list if not scheduled.
                remainingReady.push_back(taskId);
            }
        }
        // Replace ready list with those not scheduled.
        ready = remainingReady;
        // Sort ready list to maintain ordering.
        std::sort(ready.begin(), ready.end());

        // If no tasks were scheduled now, then we need to advance time.
        if (!scheduledSomething) {
            // If there are running tasks, advance time to the earliest finish.
            if (!runningTasks.empty()) {
                RunningTask finishedTask = runningTasks.top();
                runningTasks.pop();
                // Advance current time to finish time.
                current_time = finishedTask.finish_time;
                finishedCount++;

                // Free resources allocated to this task.
                for (auto& req : finishedTask.resource_requirements) {
                    int resId = req.first;
                    int qty = req.second;
                    available[resId] += qty;
                }
                // Update dependents.
                for (int dependent : dependents[finishedTask.task_id]) {
                    inDegree[dependent]--;
                    if (inDegree[dependent] == 0 && !scheduled[dependent]) {
                        ready.push_back(dependent);
                    }
                }
                // Ensure deterministic order in ready list.
                std::sort(ready.begin(), ready.end());
            }
        }
        // Also, if there are still running tasks but available ready tasks are unschedulable, we should process next finish event.
        if (scheduledSomething == false && ready.empty() && !runningTasks.empty()) {
            RunningTask finishedTask = runningTasks.top();
            runningTasks.pop();
            current_time = finishedTask.finish_time;
            finishedCount++;
            for (auto& req : finishedTask.resource_requirements) {
                int resId = req.first;
                int qty = req.second;
                available[resId] += qty;
            }
            for (int dependent : dependents[finishedTask.task_id]) {
                inDegree[dependent]--;
                if (inDegree[dependent] == 0 && !scheduled[dependent]) {
                    ready.push_back(dependent);
                }
            }
            std::sort(ready.begin(), ready.end());
        }
        // If both ready and running tasks exist and you cannot schedule additional tasks due to resource constraints,
        // then process the next finish event.
        while (!ready.empty()) {
            bool anySchedulable = false;
            for (int taskId : ready) {
                const Task& task = tasks[taskId];
                bool canSched = true;
                for (auto& req : task.resource_requirements) {
                    int resId = req.first;
                    int qty = req.second;
                    if (available[resId] < qty) {
                        canSched = false;
                        break;
                    }
                }
                if (canSched) {
                    anySchedulable = true;
                    break;
                }
            }
            if (anySchedulable) break;
            if (!runningTasks.empty()) {
                RunningTask finishedTask = runningTasks.top();
                runningTasks.pop();
                current_time = finishedTask.finish_time;
                finishedCount++;
                for (auto& req : finishedTask.resource_requirements) {
                    int resId = req.first;
                    int qty = req.second;
                    available[resId] += qty;
                }
                for (int dependent : dependents[finishedTask.task_id]) {
                    inDegree[dependent]--;
                    if (inDegree[dependent] == 0 && !scheduled[dependent]) {
                        ready.push_back(dependent);
                    }
                }
                std::sort(ready.begin(), ready.end());
            } else {
                break;
            }
        }
    }

    // The last current_time corresponds to the finish time of the last completed task.
    return current_time;
}

File: java/exercises/practice/unit_test_0f74fac8_fcea_4514_a7a1_23ff043ddcb0/task_scheduler/.meta/src/main/java/TaskScheduler.java
import java.util.*;
import java.lang.reflect.Field;

public class TaskScheduler {

    // Inner class representing a task in the scheduling simulation.
    private static class TaskNode {
        int id;
        int duration;
        List<Integer> dependencies;
        Map<String, Integer> resourceRequirements;
        List<TaskNode> dependents;
        int inDegree;
        int startTime;

        TaskNode(int id, int duration, List<Integer> dependencies, Map<String, Integer> resourceRequirements) {
            this.id = id;
            this.duration = duration;
            this.dependencies = dependencies;
            this.resourceRequirements = resourceRequirements;
            this.dependents = new ArrayList<>();
            this.inDegree = 0;
            this.startTime = -1;
        }
    }

    // Running task event for simulation
    private static class RunningTask {
        TaskNode task;
        int finishTime;

        RunningTask(TaskNode task, int finishTime) {
            this.task = task;
            this.finishTime = finishTime;
        }
    }

    /**
     * Schedules tasks given a list of task objects and available resources.
     * Each task object is expected to have the following fields:
     *  - int id;
     *  - int duration;
     *  - List<Integer> dependencies;
     *  - Map<String, Integer> resourceRequirements;
     *
     * The method returns a Map where the key is the task id and the value is its scheduled start time.
     * If a feasible schedule does not exist, an empty map is returned.
     */
    public static Map<Integer, Integer> scheduleTasks(List<?> tasks, Map<String, Integer> availableResources) {
        // Convert input tasks using reflection to our internal representation.
        Map<Integer, TaskNode> nodes = new HashMap<>();
        for (Object obj : tasks) {
            try {
                // Access the fields using reflection.
                Field idField = obj.getClass().getDeclaredField("id");
                idField.setAccessible(true);
                int id = (int) idField.get(obj);
                
                Field durationField = obj.getClass().getDeclaredField("duration");
                durationField.setAccessible(true);
                int duration = (int) durationField.get(obj);
                
                Field depsField = obj.getClass().getDeclaredField("dependencies");
                depsField.setAccessible(true);
                @SuppressWarnings("unchecked")
                List<Integer> dependencies = (List<Integer>) depsField.get(obj);
                
                Field resField = obj.getClass().getDeclaredField("resourceRequirements");
                resField.setAccessible(true);
                @SuppressWarnings("unchecked")
                Map<String, Integer> resourceRequirements = (Map<String, Integer>) resField.get(obj);

                // Check if any task has a requirement that exceeds the available resource.
                for (Map.Entry<String, Integer> req : resourceRequirements.entrySet()) {
                    int available = availableResources.getOrDefault(req.getKey(), 0);
                    if (req.getValue() > available) {
                        return new HashMap<>();
                    }
                }
                
                TaskNode node = new TaskNode(id, duration, dependencies, resourceRequirements);
                nodes.put(id, node);
            } catch (Exception e) {
                // In case of any reflection error, return empty schedule.
                return new HashMap<>();
            }
        }

        // Build dependency graph. Also check for missing dependency ids.
        for (TaskNode node : nodes.values()) {
            for (int depId : node.dependencies) {
                if (!nodes.containsKey(depId)) {
                    return new HashMap<>();
                }
                // Avoid self dependency.
                if (depId == node.id && node.dependencies.size() > 0) {
                    return new HashMap<>();
                }
                TaskNode depNode = nodes.get(depId);
                depNode.dependents.add(node);
                node.inDegree++;
            }
        }

        // Check for cycle using count of tasks processed.
        int totalTasks = nodes.size();
        int countCheck = 0;
        Queue<TaskNode> cycleCheckQueue = new LinkedList<>();
        for (TaskNode node : nodes.values()) {
            if (node.inDegree == 0) {
                cycleCheckQueue.offer(node);
            }
        }
        while (!cycleCheckQueue.isEmpty()) {
            TaskNode cur = cycleCheckQueue.poll();
            countCheck++;
            for (TaskNode dependent : cur.dependents) {
                dependent.inDegree--;
                if (dependent.inDegree == 0) {
                    cycleCheckQueue.offer(dependent);
                }
            }
        }
        if (countCheck != totalTasks) {
            // There is a cycle.
            return new HashMap<>();
        }

        // Reinitialize inDegree for simulation.
        for (TaskNode node : nodes.values()) {
            node.inDegree = node.dependencies.size();
        }

        // Simulation details:
        int currentTime = 0;
        // Used resources at the current time by running tasks.
        Map<String, Integer> usedResources = new HashMap<>();
        for (String key : availableResources.keySet()) {
            usedResources.put(key, 0);
        }
        // PriorityQueue for running tasks ordered by finish time.
        PriorityQueue<RunningTask> runningPQ = new PriorityQueue<>(Comparator.comparingInt(rt -> rt.finishTime));
        // Ready tasks PriorityQueue ordered by descending duration (longest first).
        PriorityQueue<TaskNode> readyPQ = new PriorityQueue<>((a, b) -> Integer.compare(b.duration, a.duration));

        // Initialize ready list: tasks with no dependencies.
        for (TaskNode node : nodes.values()) {
            if (node.inDegree == 0) {
                readyPQ.offer(node);
            }
        }

        // Map to store result schedule.
        Map<Integer, Integer> schedule = new HashMap<>();

        // Count scheduled tasks.
        int scheduledTasks = 0;

        while (scheduledTasks < totalTasks) {
            boolean scheduledNewTask = false;
            // Try to schedule ready tasks that satisfy resource constraints.
            List<TaskNode> deferred = new ArrayList<>();
            while (!readyPQ.isEmpty()) {
                TaskNode candidate = readyPQ.poll();
                if (canAllocate(candidate, availableResources, usedResources)) {
                    // Allocate resources for the task.
                    allocateResources(candidate, usedResources);
                    candidate.startTime = currentTime;
                    schedule.put(candidate.id, currentTime);
                    scheduledNewTask = true;
                    scheduledTasks++;
                    // Add to running tasks.
                    runningPQ.offer(new RunningTask(candidate, currentTime + candidate.duration));
                } else {
                    deferred.add(candidate);
                }
            }
            // Put back deferred tasks.
            for (TaskNode node : deferred) {
                readyPQ.offer(node);
            }

            if (!scheduledNewTask) {
                if (runningPQ.isEmpty()) {
                    // No tasks can be scheduled and none are running.
                    return new HashMap<>();
                }
                // Advance time to next finish event.
                RunningTask finishedTask = runningPQ.poll();
                int nextTime = finishedTask.finishTime;
                // Advance current time.
                currentTime = nextTime;
                // Release resources for all tasks that finish at this time.
                List<RunningTask> finishedAtTheSameTime = new ArrayList<>();
                finishedAtTheSameTime.add(finishedTask);
                while (!runningPQ.isEmpty() && runningPQ.peek().finishTime == currentTime) {
                    finishedAtTheSameTime.add(runningPQ.poll());
                }
                for (RunningTask rt : finishedAtTheSameTime) {
                    releaseResources(rt.task, usedResources);
                    // Update dependent tasks.
                    for (TaskNode dependent : rt.task.dependents) {
                        dependent.inDegree--;
                        if (dependent.inDegree == 0) {
                            readyPQ.offer(dependent);
                        }
                    }
                }
            }
        }
        return schedule;
    }

    // Helper method to check if task resources can be allocated.
    private static boolean canAllocate(TaskNode task, Map<String, Integer> available, Map<String, Integer> used) {
        for (Map.Entry<String, Integer> req : task.resourceRequirements.entrySet()) {
            int availableAmount = available.getOrDefault(req.getKey(), 0);
            int usedAmount = used.getOrDefault(req.getKey(), 0);
            if (usedAmount + req.getValue() > availableAmount) {
                return false;
            }
        }
        return true;
    }

    // Helper method to allocate resources.
    private static void allocateResources(TaskNode task, Map<String, Integer> used) {
        for (Map.Entry<String, Integer> req : task.resourceRequirements.entrySet()) {
            used.put(req.getKey(), used.getOrDefault(req.getKey(), 0) + req.getValue());
        }
    }

    // Helper method to release resources.
    private static void releaseResources(TaskNode task, Map<String, Integer> used) {
        for (Map.Entry<String, Integer> req : task.resourceRequirements.entrySet()) {
            used.put(req.getKey(), used.getOrDefault(req.getKey(), 0) - req.getValue());
        }
    }
}